nilearn/surface/surface.py

```diff
@@ -1753,3 +1779,167 @@ def from_volume(
+        SurfaceImage to check.
+
+    dim : int
+        Dimensions the data should have.
+
+    var_name : str, optional
+        Name of the variable to include in the error message.
+
+    Returns
+    -------
+    raise ValueError if the data of the SurfaceImage is not of the given
+    dimension.
+    """
+    n_dim_left = img.data.parts["left"].ndim
+    n_dim_right = img.data.parts["right"].ndim
+    if not all(x == dim for x in [n_dim_left, n_dim_right]):
+        raise ValueError(
+            f"Data for each hemisphere of {var_name} should be {dim}D, "
+            f"but found {n_dim_left}D for left and {n_dim_right}D for right."
+        )
+
+
+def get_min_max_surface_image(img):
+    """Get min and max across hemisphere for a SurfaceImage.
+
+    Parameters
+    ----------
+    img : SurfaceImage
+
+    Returns
+    -------
+    vmin : float
+
+    vmax : float
+    """
+    vmin = min(min(x.ravel()) for x in img.data.parts.values())
+    vmax = max(max(x.ravel()) for x in img.data.parts.values())
+    return vmin, vmax
+
+
+def concat_extract_surface_data_parts(img):
+    """Concatenate the data of a SurfaceImage across hemispheres and return
+    as a numpy array.
+
+    Parameters
+    ----------
+    img : :obj:`~nilearn.surface.SurfaceImage` object
+        SurfaceImage whose data to concatenate and extract.
+
+    Returns
+    -------
+    :obj:`~numpy.ndarray`
+        Concatenated data across hemispheres.
+    """
+    return np.concatenate(list(img.data.parts.values()), axis=0)
+
+
+def concat_imgs(imgs):
+    """Concatenate the data of a list or tuple of SurfaceImages.
+
+    Assumes all images have same meshes.
+
+    Parameters
+    ----------
+    imgs : :obj:`list` or :obj:`tuple` of SurfaceImage object
+
+    Returns
+    -------
+    SurfaceImage object
+    """
+    if not isinstance(imgs, (tuple, list)) or any(
+        not isinstance(x, SurfaceImage) for x in imgs
+    ):
+        raise TypeError(
+            "'imgs' must be a list or a tuple of SurfaceImage instances."
+        )
+
+    if len(imgs) == 1:
+        return imgs[0]
+
+    for i, img in enumerate(imgs):
+        check_same_n_vertices(img.mesh, imgs[0].mesh)
+        imgs[i] = at_least_2d(img)
+
+    output_data = {}
+    for part in imgs[0].data.parts:
+        tmp = [img.data.parts[part] for img in imgs]
+        output_data[part] = np.concatenate(tmp, axis=1)
+
+    return new_img_like(imgs[0], data=output_data)
+
+
+def mean_img(img):
+    """Compute mean of SurfaceImage over time points (for 'time series').
+
+    Parameters
+    ----------
+    img : SurfaceImage
+
+    Returns
+    -------
+    SurfaceImage
+    """
+    if len(img.shape) < 2 or img.shape[1] < 2:
+        data = img.data
+    else:
+        data = {
+            part: np.mean(value, axis=1).astype(float)
+            for part, value in img.data.parts.items()
+        }
+    return new_img_like(img, data=data)
+
+
+def two_dim_to_one_dim(img):
```

Comment:
I actually liked the name before, it was something like `deconcatenate...` IIRC?
But as you mentioned in the dev meeting, maybe keeping close to the Nifti image functions is better. I found that we have `iter_img` function that does a similar thing with nifti images.
https://github.com/nilearn/nilearn/blob/04be71faa15194e41e9455ca87d3b4427580ff42/nilearn/image/image.py#L726
<img width="853" alt="Screenshot 2024-12-17 at 12 04 48 PM" src="https://github.com/user-attachments/assets/ecdc4fab-24fc-4218-8676-62ff3f7f8909" />

Reply:
Could try but I was trying to align with nibabel's four_to_three but I felt that `two_to_one` was a bit strange

https://nipy.org/nibabel/reference/nibabel.funcs.html#nibabel.funcs.four_to_three

If we go for iter_img, then we probably want to implement a generator by default, no?
This is what the name implies at least.

Reply:
I checked and this is what it is returns in the end ([code](
https://github.com/nilearn/nilearn/blob/d642bb8a63e19620729fbb52d72dffb933fa707a/nilearn/_utils/niimg_conversions.py#L332))
<img width="666" alt="Screenshot 2024-12-17 at 1 25 44 PM" src="https://github.com/user-attachments/assets/48e58a05-a948-4cc9-a88b-528de7f0195d" />

the `_index_img` function just returns `new_img_like`.

So it seems like it returns a very long tuple? But either way, I think tthe end use case is the same here, no?

Reply:
na pretty sure this is an "iterator" not a tuple: it returns a generator:

```python
In [1]: foo = (a for a in [1, 2])
In [2]: type(foo)
Out[2]: generator
```